src/mango/src/mango_idx.erl

```diff
@@ -329,6 +339,87 @@ gen_name(Idx, Opts0) ->
     mango_util:enc_hex(Sha).
 
 
+get_idx_partitioned(Opts) ->
+    case proplists:get_value(partitioned, Opts) of
+        B when is_boolean(B) ->
+            B;
+        default ->
+            undefined
+    end.
+
+
+set_ddoc_partitioned(DDoc, Idx) ->
+    % We have to verify that the new index being added
+    % to this design document either matches the current
+    % ddoc's design options *or* this is a new design doc
+    #doc{
+        id = DDocId,
+        revs = Revs,
+        body = {BodyProps}
+    } = DDoc,
+    OldDOpts = couch_util:get_value(<<"options">>, BodyProps),
+    OldOpt = case OldDOpts of
+        {OldDOptProps} when is_list(OldDOptProps) ->
+            couch_util:get_value(<<"partitioned">>, OldDOptProps);
+        _ ->
+            undefined
+    end,
+    % If new matches old we're done
+    if Idx#idx.partitioned == OldOpt -> DDoc; true ->
+        % If we're creating a ddoc then we can set the options
+        case Revs == {0, []} of
+            true when Idx#idx.partitioned /= undefined ->
+                set_ddoc_partitioned_option(DDoc, Idx#idx.partitioned);
+            true when Idx#idx.partitioned == undefined ->
+                DDoc;
+            false ->
+                ?MANGO_ERROR({partitioned_option_mismatch, DDocId})
+        end
+    end.
+
+
+set_ddoc_partitioned_option(DDoc, Partitioned) ->
+    #doc{
+        body = {BodyProps}
+    } = DDoc,
+    NewProps = case couch_util:get_value(<<"options">>, BodyProps) of
```

Comment:
BodyProps is user provided input so there's no guarantee that the `partitioned` key was provided in the JSON blob.

Reply:
I made a typo in my example it should be as follows and I still think it would work:
```
{Opts} = couch_util:get_value(<<"options">>, BodyProps, {[]}),
Opt = {<<"partitioned">>, Partitioned},
New = lists:keystore(<<"partitioned">>, 1, Opts, Opt),
NewProps = lists:keystore(<<"options">>, 1, BodyProps, {<<"options">>, New}),
DDoc#doc{body = {NewProps}}.
```

- If `options` key is not provided the Opts would be `{[]}` since it is specified in default argument in the call to `couch_util:get_value/3`.
- If `partitioned` key is not provided it we would add it in lists:keystore/4.
- If `partitioned` key is provided in the `BodyProps` we inject it into `options->partitioned`
- If value for `partitioned` is not in the right format is handled elsewhere.